fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java

```diff
@@ -751,33 +782,43 @@ public void normalizeConjuncts(FragmentNormalizer normalizer, TNormalPlanNode pl
         List<Integer> remappedSlotIds = normalizer.remapSlotIds(slotIds);
 
         planNode.olap_scan_node.setRemapped_slot_ids(remappedSlotIds);
-        planNode.olap_scan_node.setSelected_column(slotIdToColNames.stream().map(c->c.second).collect(Collectors.toList()));
+        planNode.olap_scan_node.setSelected_column(
+                slotIdToColNames.stream().map(c -> c.second).collect(Collectors.toList()));
 
         List<Map.Entry<Long, Range<PartitionKey>>> rangeMap = Lists.newArrayList();
         try {
-            rangeMap = rangePartitionInfo.getSortedRangeMap(new HashSet<>(selectedPartitionIds));
+            rangeMap = rangePartitionInfo.getSortedRangeMap(selectedPartIdSet);
         } catch (AnalysisException ignored) {
         }
         conjuncts = normalizer.getPartitionRangePredicates(conjuncts, rangeMap, rangePartitionInfo, slotId);
         planNode.setConjuncts(normalizer.normalizeExprs(conjuncts));
     }
 
-    private boolean isUnCacheable() {
+    private boolean isUnCacheable(FragmentNormalizer normalizer) {
+        normalizer.setKeysType(olapTable.getKeysType());
         switch (olapTable.getKeysType()) {
             case DUP_KEYS:
+                normalizer.setCanUseMultiVersion(true);
+                return false;
+            case PRIMARY_KEYS:
+            case UNIQUE_KEYS:
+                normalizer.setCanUseMultiVersion(false);
+                return false;
             case AGG_KEYS: {
                 List<Column> columns = selectedIndexId == -1 ? olapTable.getBaseSchema() :
                         olapTable.getSchemaByIndexId(selectedIndexId);
-                return columns.stream().anyMatch(
+                boolean hasReplaceAgg = columns.stream().anyMatch(
                         c -> c.isAggregated() && c.getAggregationType().isReplaceFamily());
+                normalizer.setCanUseMultiVersion(!hasReplaceAgg);
```

Comment:
What's the behavior when the cache is uncacheable and canUseMultiVersion?

Reply:
uncacheable is confusing,  so it is renamed to setIfCanUseMultiVersionCache. the behavior is commented